### PR TITLE
doc: adding database transaction diagram and explanation

### PR DESCRIPTION
**Pull request by m1sterc001guy**
This PR adds a diagram and explanation of the database transaction interfaces/structs and their purposes.

Issue: https://github.com/fedimint/fedimint/issues/1831

Link to branch: https://github.com/m1sterc001guy/fedimint/blob/db_tx_doc/docs/database.md

### CODE REVIEW

**Review by Maan2003 — COMMENTED** [review]
Thanks this is awesome.

Just some nits bellow.

File: docs/database.md [hunk 1]
```diff
@@ -104,3 +104,118 @@ The client is not currently modularized, so nothing is prepended to the below pr
...
+      new_module_tx()
+      get_value()
+      find_by_prefix()
+      insert_entry()
+      insert_new_entry()
+      add_notification_key()
+      remove_entry()
+      remove_by_prefix()
+      rollback_tx_to_savepoint()
+      set_tx_savepoint()
```
[the remark this answers is not in the document]
Comment (m1sterc001guy) [reply docs/database.md:173]: I wanted to show that `ModuleDatabaseTransaction` cannot commit. So I'll add that one too. I can remove the rest.

File: docs/database.md [hunk 2]
```diff
@@ -104,3 +104,118 @@ The client is not currently modularized, so nothing is prepended to the below pr
...
+        Transaction optimistic_tx
+    }
+    class MemTransaction{
+        <<interface IDatabaseTransaction>>
+        Vec operations
+        BTreeMap tx_data
+        &Database db
+        BTreeMap savepoint
+        int num_pending_operations
+        int num_savepoint_operations
```
[the remark this answers is not in the document]
Comment (m1sterc001guy) [reply docs/database.md:216]: Can remove, its' probably too much detail

File: docs/database.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (m1sterc001guy) [reply docs/database.md:173]: `with_module_prefix` and `new_module_tx` are useful too for showing how the sub-types are created

**Review by dpc — APPROVED** [review]

**Review by Maan2003 — APPROVED** [review]

**Comment by justinmoon** [comment]
That diagram is lovely

**Pull request merged**